### ISSUE TITLE
gio::Application::get_default() is unsafe

### ISSUE THREAD

**Issue by sdroege** [gtk-rs/gtk-rs-core#44]
It allows retrieving the application from any thread, not just the main thread (or more general the thread where the application was created). `gio::Application` is not thread-safe, so this is a problem.

**Comment by jf2048** [comment]
It seems this can never be completely safe because `g_application_set_default` is not thread safe, and that gets called internally by `GApplication` implementations. I guess what that means is, a `GApplication` can only be referenced in the main thread, but we don't have any way of knowing what the main thread really is, getting the current thread id is just a guess.

It could be made a little more robust by checking if we can acquire the main context before storing the `qdata`, but that still isn't completely safe in all cases because it only checks for a running application...

**Comment by sdroege** [comment]
Yeah the problem with making `get_default()` unsafe is that it's a very useful API to have in GTK applications and requiring those to use unsafe code is not ideal (`set_default()` is fine IMHO, who needs that anyway...)

**Comment by jf2048** [comment]
That will not fully make it sound but will probably correct most of the issues people have. The wrapper macro will need some modifications here to do the thread checking so we may need to see https://github.com/gtk-rs/gtk-rs-core/issues/916 for that

**Comment by sdroege** [comment]
I think the easiest would be to implement `default()` manually for now and implement the same assertion that is also used in e.g. `glib::MainContext::spawn_local()`. That should cover the easiest mistakes at least.

More correct would be to forbid construction of a `gio::Application` from another thread but that requires `glib::wrapper!` macro changes as @jf2048 says, and not very easy ones.

Does someone want to do the easy part for now?

**Comment by jf2048** [comment]
> I think the easiest would be to implement `default()` manually for now and implement the same assertion that is also used in e.g. `glib::MainContext::spawn_local()`. That should cover the easiest mistakes at least.

How can we do that, since `Application` does not reference a `MainContext`? I think any solution will need to modify the wrapper and do the `qdata` solution described above.

**Comment by sdroege** [comment]
Can you actually use an `Application` on a non-default main context?
